node/src/components/consensus/era_supervisor.rs

```diff
@@ -463,6 +501,79 @@ where
     pub(crate) fn is_initialized(&self) -> bool {
         self.is_initialized
     }
+
+    fn handle_initialize_eras(
+        &mut self,
+        key_blocks: HashMap<EraId, BlockHeader>,
+        mut maybe_validators: Option<BTreeMap<PublicKey, U512>>,
+        timestamp: Timestamp,
+        genesis_start_time: Timestamp,
+    ) -> HashMap<EraId, ProtocolOutcomes<I, ClContext>> {
+        let mut result_map = HashMap::new();
+
+        for era_id in self.iter_past(self.current_era, self.bonded_eras * 2) {
+            let newly_slashed;
+            let validators;
+            let start_height;
+            let era_start_time;
+
+            if era_id.is_genesis() {
+                newly_slashed = vec![];
+                validators = maybe_validators
+                    .take()
+                    .expect("missing genesis era validators");
+                start_height = 0;
+                era_start_time = genesis_start_time;
+            } else {
+                let key_block = key_blocks.get(&era_id).expect("missing key block");
+                start_height = key_block.height() + 1;
+                era_start_time = key_block.timestamp();
+                if era_id == self.protocol_config.last_activation_point {
+                    newly_slashed = vec![];
+                    validators = maybe_validators
+                        .take()
+                        .expect("missing activation era validators");
+                } else {
+                    newly_slashed = key_block
+                        .era_end()
+                        .expect("key block must be a switch block")
+                        .equivocators
+                        .clone();
+                    validators = key_block
+                        .next_era_validator_weights()
+                        .expect("missing validators from key block")
+                        .clone();
+                }
+            }
+
+            let slashed = self
+                .iter_past(era_id, self.bonded_eras)
+                .filter_map(|old_id| key_blocks.get(&old_id).and_then(|bhdr| bhdr.era_end()))
+                .flat_map(|era_end| era_end.equivocators.clone())
+                .collect();
+
+            let booking_era_id = era_id.saturating_sub(self.protocol_config.auction_delay);
+            let seed = key_blocks.get(&booking_era_id).map_or(0, |booking_block| {
+                Self::era_seed(booking_block.hash(), booking_block.accumulated_seed())
+            });
+
+            let results = self.new_era(
+                era_id,
+                timestamp,
+                validators,
+                newly_slashed,
+                slashed,
+                seed,
+                era_start_time,
+                start_height,
+            );
+            result_map.insert(era_id, results);
+        }
+
+        self.is_initialized = true;
+        self.next_block_height = self.active_eras[&self.current_era].start_height;
+        result_map
+    }
```

Comment:
Do I have to review this method or is it simply the old code extracted?

Reply:
It has changed a bit, yes. The old version did use pre-restart switch blocks.